src/enclave/Enclave/FlatbuffersWriters.cpp

```diff
@@ -75,30 +85,142 @@ void RowWriter::maybe_finish_block() {
+    uint8_t mac_lst[num_macs * SGX_AESGCM_MAC_SIZE];
+    uint8_t global_mac[OE_HMAC_SIZE];
+    EnclaveContext::getInstance().hmac_mac_lst(mac_lst, global_mac);
+
+    int curr_pid = EnclaveContext::getInstance().get_pid();
+    char* untrusted_curr_ecall_str = oe_host_strndup(curr_ecall.c_str(), curr_ecall.length());
+
+    // Copy mac list to untrusted memory
+    uint8_t* untrusted_mac_lst = nullptr;
+    ocall_malloc(num_macs * SGX_AESGCM_MAC_SIZE, &untrusted_mac_lst);
+    std::unique_ptr<uint8_t, decltype(&ocall_free)> mac_lst_ptr(untrusted_mac_lst, 
+        &ocall_free);
+    memcpy(mac_lst_ptr.get(), mac_lst, num_macs * SGX_AESGCM_MAC_SIZE);
+
+    // Copy global mac to untrusted memory
+    uint8_t* untrusted_global_mac = nullptr;
+    ocall_malloc(OE_HMAC_SIZE, &untrusted_global_mac);
+    std::unique_ptr<uint8_t, decltype(&ocall_free)> global_mac_ptr(untrusted_global_mac, 
+        &ocall_free);
+    memcpy(global_mac_ptr.get(), global_mac, OE_HMAC_SIZE);
+
+    // This is an offset into enc block builder
+    auto log_entry_serialized = tuix::CreateLogEntry(enc_block_builder,
+        enc_block_builder.CreateString(std::string(untrusted_curr_ecall_str)),
+        curr_pid,
+        -1, // -1 for not yet set rcv_pid
+        job_id,
+        num_macs,
+        enc_block_builder.CreateVector(mac_lst_ptr.get(), num_macs * SGX_AESGCM_MAC_SIZE),
+        enc_block_builder.CreateVector(global_mac_ptr.get(), OE_HMAC_SIZE));
+
+    curr_log_entry_vector.push_back(log_entry_serialized);
+
+    std::vector<LogEntry> past_log_entries = EnclaveContext::getInstance().get_ecall_log_entries();
+
+    for (LogEntry le : past_log_entries) {
+      char* untrusted_ecall_op_str = oe_host_strndup(le.ecall.c_str(), le.ecall.length());
+      auto past_log_entry_serialized = tuix::CreateLogEntry(enc_block_builder,
+          enc_block_builder.CreateString(std::string(untrusted_ecall_op_str)),
+          le.snd_pid,
+          le.rcv_pid,
+          le.job_id);
+      past_log_entries_vector.push_back(past_log_entry_serialized);
+    }
+
+    num_past_log_entries.push_back(past_log_entries.size());
+   
+    // We will MAC over global_mac || curr_ecall || snd_pid || rcv_pid || job_id || num_macs 
+    // || global_mac || num past log entries || past log entries
+    int num_past_entries = (int) past_log_entries.size();
+    int past_ecalls_lengths = get_past_ecalls_lengths(past_log_entries, 0, num_past_entries);
+
+    // curr_log_entry contains:
+    //  * string curr_ecall of size curr_ecall.length()
+    //  * global_mac of size OE_HMAC_SIZE
+    //  * 5 ints
+    // 1 past log entry contains:
+    //  * string ecall of size ecall.length()
+    //  * 3 ints
+    int num_bytes_to_mac = OE_HMAC_SIZE + 5 * sizeof(int) + curr_ecall.length() * sizeof(char) 
+      + 3 * sizeof(int) * past_log_entries.size() + past_ecalls_lengths * sizeof(char);
+    uint8_t to_mac[num_bytes_to_mac];
+
+    uint8_t hmac[32];
```

Comment:
Isn't this more of a global MAC than the current global MAC in LogEntry? Maybe you can rename the `global_mac` to `log_mac`?

Reply:
I've renamed this to `mac_lst_mac` as it's not exactly a MAC over the log entry.